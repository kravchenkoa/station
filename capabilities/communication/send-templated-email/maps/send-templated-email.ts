/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable jest/no-try-expect */
/* eslint-disable jest/no-conditional-expect */
/* eslint-disable jest/no-export */

import { MappedHTTPError, Provider } from '@superfaceai/one-sdk';

import {
  CommunicationSendTemplatedEmailProfile,
  SuperfaceClient,
} from '../../../../superface/sdk';

export const sendTemplatedEmailTest = (providerName: string): void => {
  describe(`communication/send-templated-email/${providerName}`, () => {
    let client: InstanceType<typeof SuperfaceClient>;
    let profile: CommunicationSendTemplatedEmailProfile;
    let provider: Provider;

    beforeEach(async () => {
      client = new SuperfaceClient();
      profile = await client.getProfile('communication/send-templated-email');
      provider = await client.getProvider(providerName);
    });

    it('should find provider', async () => {
      expect(provider).not.toBeUndefined();
    });

    describe('SendTemplatedEmail', () => {
      describe('when all inputs are correct', () => {
        it('should return messagaId as result', async () => {
          const result = await profile.useCases.SendTemplatedEmail.perform(
            {
              from:
                process.env[
                  `${providerName.toUpperCase()}_COMMUNICATION_SENDEMAIL_FROM`
                ],
              to: process.env.COMMUNICATION_SENDEMAIL_TO,
              templateId:
                process.env[
                  `${providerName.toUpperCase()}_COMMUNICATION_SENDEMAIL_TEMPLATE_ID`
                ],
              templateData: {
                from:
                  process.env[
                    `${providerName.toUpperCase()}_COMMUNICATION_SENDEMAIL_FROM`
                  ],
                to: process.env.COMMUNICATION_SENDEMAIL_TO,
              },
            },
            { provider }
          );

          expect(typeof result.unwrap().messageId).toBe('string');
        });
      });

      describe('when inputs are invalid', () => {
        it('should throw error on unwrap', async () => {
          const result = await profile.useCases.SendTemplatedEmail.perform(
            {
              to: 'invalidemail',
              from: 'invalidemail',
              templateId: 'invalid',
              templateData: null,
            },
            { provider }
          );

          try {
            result.unwrap();
            throw new Error('must throw something');
          } catch (error) {
            expect(error).toBeInstanceOf(MappedHTTPError);
            expect(error.properties.title).toBe('Invalid inputs');
          }
        });
      });
    });
  });
};
