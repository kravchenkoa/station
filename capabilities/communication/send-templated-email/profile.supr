"""
Send Templated Email
Send templated transactional email to one recipient.
"""

name = "communication/send-templated-email"
version = "1.0.0"

"""
Send Templated Email
Send templated transactional email to one recipient.
Requires template defined on provider side. 
"""
usecase SendTemplatedEmail unsafe {
  input {
    """
    From
    The sender's email address.
    """
    from!

    """
    To
    The recipient's email address.
    """
    to!

    """
    Template Identifier
    The template to use when sending email message.
    """
    templateId!

    """
    Template Data
    Template data to be applied to the specified template to generate html, test, and subject.
    The value is a collection of key/value pairs following the pattern `variable_name`: `value to insert`.
    This field should be used in combination with `templateId` to identify what template to use.
    """
    templateData!
  }

  result {
    """
    Message Identifier
    The identifier is provider-specific and not unique.
    """
    messageId!
  }

  error {
    """
    Title
    A short, human-readable summary of the problem type.
    """
    title!

    """
    A human-readable explanation specific to this occurrence of the problem.
    """
    detail
  }
}
