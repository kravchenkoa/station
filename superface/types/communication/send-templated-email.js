"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.communicationSendTemplatedEmail = void 0;
var one_sdk_1 = require("@superfaceai/one-sdk");
var profile = {
    /**
     * Send Templated Email
     * Send templated transactional email to one recipient.
     * Requires template defined on provider side.
     **/
    "SendTemplatedEmail": one_sdk_1.typeHelper()
};
exports.communicationSendTemplatedEmail = {
    "communication/send-templated-email": profile
};
