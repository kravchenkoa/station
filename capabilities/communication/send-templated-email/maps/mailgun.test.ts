import { sendTemplatedEmailTest } from './send-templated-email';

sendTemplatedEmailTest('mailgun', {
  from: 'hello@demo.superface.org',
  to: 'hello@superface.ai',
  domain: 'demo.superface.org',
  templateId: 'station-test',
});
