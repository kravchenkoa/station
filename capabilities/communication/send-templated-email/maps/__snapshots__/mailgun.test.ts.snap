// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`communication/send-templated-email/mailgun SendTemplatedEmail when all inputs are correct should return messageId as result 1`] = `
Object {
  "value": Object {
    "messageId": "<20210924083001.1.87AA7862F3410AB2@demo.superface.org>",
  },
}
`;

exports[`communication/send-templated-email/mailgun SendTemplatedEmail when inputs are invalid should throw error 1`] = `
Object {
  "error": "HTTPError: Expected HTTP error
AST Path: definitions[0].statements[1].responseHandlers[1].statements[0]
Original Map Location: Line 30, column 7",
}
`;
